src/Tracepoints.jl

```diff
@@ -0,0 +1,497 @@
+    argspec = parse_args(mod, args)
+    spec = create_tracepoint_spec(source, argspec, args; alias_with)
+    register_probe(mod, category, kind, spec)
+    return create_tracepoint_call(category, kind, lib_id, argspec, spec)
+end
+# Semaphore and probe payload not yet loaded
+function create_tracepoint_call(category, kind, lib_id, argspec, spec)
+    abi_type, args_box, arg_name = wrap_tracepoint_args(argspec)
+    return quote
+        $probe_maybe_trigger($spec, $(QuoteNode(category)), $(QuoteNode(kind)), $lib_id, $(Val{abi_type}()), $args_box, $(QuoteNode(arg_name)))
+    end
+end
+# Semaphore and probe payload already loaded
+function create_tracepoint_call_loaded(category, kind, lib_id, argspec, spec, sema_set, payload)
+    abi_type, args_box, arg_name = wrap_tracepoint_args(argspec)
+    return quote
+        if $sema_set && $payload != C_NULL
+            $probe_trigger($spec, $(QuoteNode(category)), $(QuoteNode(kind)), $lib_id, $(Val{abi_type}()), $args_box, $(QuoteNode(arg_name)))
+        end
+    end
+end
+
+macro tracepoint(kind::QuoteNode, category::QuoteNode, lib_id, args=Expr(:tuple))
+    create_tracepoint!(__module__, __source__, category.value, kind.value, lib_id, args)
+end
+macro region_start(category::QuoteNode, args=Expr(:tuple))
+    create_tracepoint!(__module__, __source__, category.value, :start, 0, args)
+end
+macro region_finish(category::QuoteNode, lib_id, args=Expr(:tuple))
+    create_tracepoint!(__module__, __source__, category.value, :finish, lib_id, args)
+end
+
+"""
+    @region :category arg1 arg2... ex
+
+Constructs a pair of tracepoints around the execution of `ex`. The start
+tracepoint executes just before `ex`, and the finish tracepoint executes just
+after `ex`. The category for both tracepoints is `:category`, and must be a
+literal `Symbol` (as the tracepoints are constructed while `@region` is being
+parsed). The kind for each tracepoint is `:start` and `:finish` respectively
+(used when programming or enabling/disabling tracepoints).
+
+These tracepoints start out disabled and unprogrammed, and by default will do
+nothing (having negligible performance impact, if any). They can be programmed
+with [`set!`](@ref), and enabled with [`enable!`](@ref).
+
+Arguments may be provided in a few formats:
+
+`arg::T` - This specifies that the argument with name `arg` has type `T`, where
+`T` is a concrete type. In this form, `arg` must be a variable in scope that
+`@region` can read during start and finish tracepoint execution. If `arg` is
+modified during the execution of `ex`, its new value will be passed into the
+finish tracepoint.
+
+`arg::T=value` - This format is similar to `arg::T`, except that `arg` is not
+read from a variable with the same name, but is instead hard-coded as `value`
+for both start and finish tracepoints.
+
+`(start_arg::ST, finish_arg::FT)` - This format is like `arg::T`, except that
+it allows specifying different values for the start and finish tracepoints.
```

Comment:
Do you have a particular use case in mind for this? Additionally, when is this evaluated and where does the name(s) come from?

This feels a bit ambiguous-in-spirit with explicit tuple initializers that happen to include type assertions like
```
@region z::Any=(1::Int64, 0::Int64)
```

Reply:
It's convenient to have this for start-finish pairs in functions where the argument points to two different values at start and finish - the extra type specification is there since those values may also change types in a predictable manner. The alternative is to have to use `@region_start/finish`, which I prefer not to use since it's easy to mess them up. Do you have another recommendation for syntax for this kind of need?

> `@region z::Any=(1::Int64, 0::Int64)`

The macro should really only allow `@region z=(1::Int64, 0::Int64)` (although I haven't tested that).